src/ratelimit_action.rs

```diff
@@ -0,0 +1,332 @@
+                    _ => panic!("Only scalar values can be sent as data"),
+                },
+                Err(err) => {
+                    error!("Failed to evaluate {:?}: {}", self.expression, err);
+                    panic!("Err out of this!")
+                }
+            },
+        );
+        let mut descriptor_entry = RateLimitDescriptor_Entry::new();
+        descriptor_entry.set_key(key);
+        descriptor_entry.set_value(value);
+        descriptor_entry
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ConditionalData {
+    pub data: OnceCell<Vec<DescriptorEntryBuilder>>,
+    pub predicates: OnceCell<Vec<Predicate>>,
+}
+
+impl ConditionalData {
+    pub fn new(action: &Action) -> Result<Self, String> {
+        let mut predicates = Vec::default();
+        for predicate in &action.predicates {
+            predicates.push(Predicate::new(predicate).map_err(|e| e.to_string())?);
+        }
+
+        let mut data_exp_list = Vec::default();
+        for datum in &action.data {
+            data_exp_list.push(DescriptorEntryBuilder::new(&datum.item)?);
+        }
+        let conditional_data = ConditionalData {
+            data: OnceCell::new(),
+            predicates: OnceCell::new(),
+        };
+        conditional_data
+            .predicates
+            .set(predicates)
+            .expect("Predicates must not be compiled yet!");
+
+        conditional_data
+            .data
+            .set(data_exp_list)
+            .expect("Data expressions must not be compiled yet!");
+        Ok(conditional_data)
+    }
+
+    fn predicates_apply(&self) -> bool {
+        let predicates = self
+            .predicates
+            .get()
+            .expect("predicates must be compiled by now");
+        predicates.is_empty()
+            || predicates.iter().all(|predicate| match predicate.test() {
+                Ok(b) => b,
+                Err(err) => {
+                    error!("Failed to evaluate {:?}: {}", predicates, err);
+                    panic!("Err out of this!")
+                }
+            })
+    }
+
+    pub fn entries(&self) -> RepeatedField<RateLimitDescriptor_Entry> {
+        if !self.predicates_apply() {
+            return RepeatedField::default();
+        }
+
+        let mut entries = RepeatedField::default();
+        for entry_builder in self
+            .data
+            .get()
+            .expect("data must be compiled by now")
+            .iter()
+        {
+            entries.push(entry_builder.evaluate());
+        }
+
+        entries
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RateLimitAction {
+    grpc_service: Rc<GrpcService>,
+    scope: String,
+    service_name: String,
+    conditional_data_sets: Vec<ConditionalData>,
+}
+
+impl RateLimitAction {
+    pub fn new(action: &Action, service: &Service) -> Result<Self, String> {
+        Ok(Self {
+            grpc_service: Rc::new(GrpcService::new(Rc::new(service.clone()))),
```

Comment:
So that's an `Rc` of ... one _always_?

Reply:
I will take a look

Reply:
I think similar to my previous comment, we also used to create all the `GrpcService` up front and share them as we know these won't change, now we're still creating them up front on configure time but one for every action?

Reply:
Sure, I think it's fine for now, but I personally think it's wrong to have as many `GrpcService`s as we have actions, with 2 services defined in the config and N actions (assuming no merges) we'll have N services when we had 2 before.